specs/hierarchical_consensus.md

```diff
@@ -0,0 +1,1082 @@
+
+The Subnet Actor is the public contract accessible by users in the system to determine the kind of child subnet being spawned and controlled by the actor. From the moment the SA for a new subnet is spawned in the parent chain, users looking to participate in the subnet can instantiate their new chain and even start mining on it. However, in order for the subnet to be able to interact with the rest of the hierarchy, it needs to be registered by staking an amount of native tokens over the `CollateralThreshold` in the parent's SCA _(see [Collateral and slashing](#Collateral-and-slashing "Collateral and slashing"))_.
+
+##### Subnet Actor Interface
+A reference implementation of the subnet actor interface exists, but users are permitted to implement custom governing policies and mechanics that better suit their needs (e.g. request a dynamic collateral for validators, add a leaving fee coefficient to penalize validators leaving the subnet before some time frame, require a minimum number of validators, include latency limits, etc.).
+
+> Interface every subnet actor needs to implement. These functions
+> are triggered when a `message` is sent for their corresponding `methodNum`. 
+```go
+type SubnetActor interface{
+    // Initializes the state of the subnet actor and sets
+    // all its initial parameters.
+    // (methodNum = 1)
+    Constructor(ConstructParams)
+    
+    // It expects as `value` the amount of collateral the 
+    // source wants to stake to become part of the subne.
+    // It triggers all check to determine if the source is eligible
+    // to become part of the subnet, and must propagate an `AddCollateral` message
+    // to the SCA to trigger the registration.
+    // 
+    // Join is also responsible to send a `Register` message to the SCA
+    // whenever the `CollateralThreshold` is reached. This function can also
+    // be used to AddCollateral for the subnet in the SCA. Additional
+    // checks for this operation may be included. As a result of the
+    // execution an `AddCollateral` message must be sent to the SCA 
+    // including the added collateral.
+    // (methodNum = 2)
+    Join()
+    
+    // Called by participants that want to leave the subnet. It triggers all
+    // the checks to see if the source is eligible to leave the subnet.
+    // This function must propagate a `ReleaseCollateral` message to the SCA.
+    // The SCA will then release the corresponding stake of the validator in
+    // a message to the subnet actor so the subnet actor can trigger a message
+    // returning the funds to the leaving validtor.
+    // (methodNum = 3)
+    Leave()
+    
+    // Kill performs all the sanity-checks required before completely
```

Comment:
It might be helpful giving an example of what kind of checks the reference implementation does, like there are in the `Checkpoint` method.